README.md
# node-watershed

A simple implementation of WebSockets for use with node.js 'http'-style web
server or client.  Deals only with [RFC6455][1], ignoring any Browser-specific
peculiarities, curiosities, and fallback mechanisms.

## API

### Watershed

The core of this library is a Factory-style class `Watershed`.  It has several
static methods:

#### Watershed.accept(http.ServerRequest, net.Socket, Buffer[, Boolean[, String[] ]])

Responds to a client's request to Upgrade an HTTP connection to a WebSocket and
returns a `WatershedConnection`, which is also an `EventEmitter`.

The arguments to this method should ideally come from the `'upgrade'` event on
a node.js `http.Server`.  For example:

```javascript
var shed = new Watershed();
var srv = http.createServer();
srv.listen(8080);
srv.on('upgrade', function(req, socket, head) {
        var wsc;
        try {
                wsc = shed.accept(req, socket, head);
        } catch (ex) {
                console.error('error: ' + ex.message);
                return socket.end();
        }
        wsc.on('text', function(text) {
                console.log('received text: ' + text);
        });
        wsc.on('end', function() {
                console.log('end!');
        });
        /* ... etc ... */
});
```

The additional penultimate boolean argument may be used when implementing a
websocket-to-websocket proxy.  If set to `true`, then `accept()` returns the
raw underlying `Socket` of the connection and does not construct a
`WatershedConnection` instance.  No event handlers will be placed on the
socket, meaning you are free to continue using it directly (e.g. calling
`pipe()` on it to join it to a backend socket).

The final list-of-strings argument can be optionally provided to allow for the
use of WebSocket subprotocol negotiation.  It is a list of case-sensitive
string names of supported subprotocols.  If given, and protocol negotiation
takes place, the chosen subprotocol can be found with the `getProtocol()`
method.

#### Watershed.generateKey()

Returns a random, Base64-encoded 16-byte value suitable for use as the
`Sec-WebSocket-Key` header on an Upgrade request.  See Example usage in
`connect()`.

#### Watershed.connect(http.ClientResponse, net.Socket, Buffer, String)

Attaches a new client-side `WatershedConnection` to this presently Upgraded
socket.

The arguments to this method should ideally come from the `'upgrade'` event on
a node.js `http.Client`.  For example:

```javascript
var shed = new Watershed();
var wskey = ws.generateKey();
var options = {
    port: 8082,
    hostname: '127.0.0.1',
    headers: {
        'connection': 'upgrade',
        'upgrade': 'websocket',
        'Sec-WebSocket-Key': wskey,
        'Sec-WebSocket-Version': '13'
    }
};
var req = http.request(options);
req.end();
req.on('upgrade', function(res, socket, head) {
        var wsc = shed.connect(res, socket, head, wskey);
        wsc.send('Hi there!');
        wsc.on('end', function() {
                console.log('end!');
        });
        /* ... etc ... */
});
```

### WatershedConnection

#### Event: 'error'

Emitted once when an error occurs during processing.  The socket will be closed
and an `'end'` event will follow.  The only argument will be an instance of
`Error`.

#### Event: 'connectionReset'

Emitted when the remote peer closes the connection without sending us a CLOSE
frame.  An `'end'` event will follow.

#### Event: 'end' (code, reason)

Emitted once when the socket is closing.  If we received a graceful CLOSE
frame from the remote server, we will attempt to process it and pass `code`
and `reason` -- both of type `String`.

#### Event: 'text'

Emitted for each inbound TEXT frame.  The only argument will be a `String`
containing the UTF-8 string payload.

#### Event: 'binary'

Emitted for each inbound BINARY frame.  The only argument will be a `Buffer`
containing the binary payload.

#### Event: 'ping'

Emitted for each inbound PING frame.  The only argument will be a `Buffer`
containing the nonce in the ping request.  Note that the library presently
responds with a PONG frame for each inbound PING frame.

#### Event: 'pong'

Emitted for each inbound PONG frame.  The only argument will be a `Buffer`
containing the nonce in the pong response.

#### WatershedConnection.getProtocol()

Returns the negotiated subprotocol, if any, for this connection, as a String.
If no subprotocol negotiation took place, this method returns `null`.

#### WatershedConnection.send(data)

Sends a frame through the socket.  The single argument `data` may be a
`Buffer`, in which case a BINARY frame is sent; or a `String`, in which case a
TEXT frame is sent.

#### WatershedConnection.end(reason)

Closes the connection.  The RFC allows a reason for closing the connection to
be send in the CLOSE frame, though this is optional.  If passed, `reason`
should be a `String`.

#### WatershedConnection.destroy()

Immediately destroy the underlying socket, without sending a CLOSE frame.  You
generally want `end()`, rather than this, in order to perform RFC-compliant
connection shutdowns.

## License

MIT.

[1]: http://tools.ietf.org/html/rfc6455


CHANGES.md
# node-watershed changelog

## not yet released

## 0.4.0

* Support for Websocket sub-protocol negotiation
* Documented `detached`/raw socket argument to `accept()`
* Change to using `crypto.randomBytes()` for entropy

## 0.3.4

* Receiving a ping would crash the WatershedConnection while trying to
  reply.
* Update to latest dtrace-provider

## 0.3.3

* Handle write-after-end errors when an attempt is made to write to a
  websocket that cannot be written to.

## 0.3.2

* Initial move to Joyent org plus make check


test/basic.js
/*
 * Copyright (c) 2017, Joyent, Inc.
 */

var http = require('http');
var test = require('tape');

var Watershed = require('../lib/watershed').Watershed;
var shed = new Watershed();
var wskey = shed.generateKey();

var CLIENT;
var HTTP;
var SERVER;
var SERVER_ERR;

function failOnMessage(msg) {
	throw new Error('received unexpected message: ' + msg.toString());
}

function failOnEnd(code, reason) {
	throw new Error('connection unexpectedly ended: ' + code);
}


// --- Tests

test('setup server', function (t) {
	HTTP = http.createServer();
	HTTP.listen(9554);
	HTTP.on('upgrade', function (req, socket, head) {
		try {
			SERVER = shed.accept(req, socket, head, false,
			    ['test1', 'test2']);
		} catch (err) {
			socket.end('HTTP/1.1 400 Bad Request\r\n' +
			    'Connection: close\r\n\r\n');
			return;
		}
		SERVER.on('text', failOnMessage);
		SERVER.on('binary', failOnMessage);
		SERVER.on('end', failOnEnd);
	});
	t.end();
});

test('setup client', function (t) {
	var options = {
		port: 9554,
		hostname: '127.0.0.1',
		headers: {
			'connection': 'upgrade',
			'upgrade': 'websocket',
			'Sec-WebSocket-Key': wskey,
			'Sec-WebSocket-Version': 13
		}
	};
	var req = http.request(options);
	req.end();
	req.on('upgrade', function (res, socket, head) {
		socket.setNoDelay(true);
		CLIENT = shed.connect(res, socket, head, wskey);
		t.strictEqual(CLIENT.getProtocol(), null);
		t.strictEqual(SERVER.getProtocol(), null);
		CLIENT.on('text', failOnMessage);
		CLIENT.on('binary', failOnMessage);
		CLIENT.on('end', failOnEnd);
		t.end();
	});
});

test('server pings client', function (t) {
	var pingRecvd = false;

	CLIENT.once('ping', function () {
		pingRecvd = true;
	});

	SERVER.once('pong', function () {
		t.ok(pingRecvd, 'received ping');
		t.end();
	});

	SERVER._ws_writePing(new Buffer(0));
});

test('client pings server', function (t) {
	var pingRecvd = false;

	SERVER.once('ping', function () {
		pingRecvd = true;
	});

	CLIENT.once('pong', function () {
		t.ok(pingRecvd, 'received ping');
		t.end();
	});

	CLIENT._ws_writePing(new Buffer(0));
});


test('client sends TEXT', function (t) {
	SERVER.removeListener('text', failOnMessage);
	SERVER.once('text', function (txt) {
		SERVER.on('text', failOnMessage);
		t.equal(txt, 'hello', 'correct message');
		t.end();
	});

	CLIENT.send('hello');
});

test('server sends TEXT', function (t) {
	CLIENT.removeListener('text', failOnMessage);
	CLIENT.once('text', function (txt) {
		CLIENT.on('text', failOnMessage);
		t.equal(txt, 'hello', 'correct message');
		t.end();
	});

	SERVER.send('hello');
});

test('client sends BINARY', function (t) {
	SERVER.removeListener('binary', failOnMessage);
	SERVER.once('binary', function (binary) {
		SERVER.on('binary', failOnMessage);
		t.ok(binary instanceof Buffer, 'Buffer returned');
		t.equal(binary.toString('utf-8'), 'hello', 'correct message');
		t.end();
	});

	CLIENT.send(new Buffer('hello'));
});

test('server sends BINARY', function (t) {
	CLIENT.removeListener('binary', failOnMessage);
	CLIENT.once('binary', function (binary) {
		CLIENT.on('binary', failOnMessage);
		t.ok(binary instanceof Buffer, 'Buffer returned');
		t.equal(binary.toString('utf-8'), 'hello', 'correct message');
		t.end();
	});

	SERVER.send(new Buffer('hello'));
});

test('teardown connection', function (t) {
	CLIENT.removeListener('end', failOnEnd);
	SERVER.removeListener('end', failOnEnd);

	CLIENT.on('end', function (code, reason) {
		t.equal(code, 'NORMAL', 'normal close');
		t.equal(reason, 'test ended', 'server sent reason');
		t.end();
	});

	SERVER.end('test ended');

	CLIENT = null;
	SERVER = null;
});

test('negotiate supported subprotocol', function (t) {
	var options = {
		port: 9554,
		hostname: '127.0.0.1',
		headers: {
			'connection': 'upgrade',
			'upgrade': 'websocket',
			'Sec-WebSocket-Key': wskey,
			'Sec-WebSocket-Version': 13,
			'Sec-WebSocket-Protocol': 'foobar, test1, test2'
		}
	};
	var req = http.request(options);
	req.end();
	req.on('upgrade', function (res, socket, head) {
		socket.setNoDelay(true);
		var client = shed.connect(res, socket, head, wskey);
		t.strictEqual(client.getProtocol(), 'test1');
		t.strictEqual(SERVER.getProtocol(), 'test1');

		SERVER.removeListener('end', failOnEnd);
		SERVER.once('end', function () {
			t.end();
		});
		client.end('done');
	});
});

test('negotiation failure', function (t) {
	var options = {
		port: 9554,
		hostname: '127.0.0.1',
		headers: {
			'connection': 'upgrade',
			'upgrade': 'websocket',
			'Sec-WebSocket-Key': wskey,
			'Sec-WebSocket-Version': 13,
			'Sec-WebSocket-Protocol': 'foobar, aaaa'
		}
	};
	var req = http.request(options);
	req.end();
	req.on('upgrade', function (res, socket, head) {
		t.fail();
	});
	req.on('response', function (res) {
		t.equal(res.statusCode, 400);
		t.end();
	});
});

test('teardown server', function (t) {
	HTTP.close();
	t.end();
});


test/basic/client.js
#!/usr/bin/env node
/*
 * Copyright 2016 Joyent, Inc.
 */

var http = require('http');
var Watershed = require('../../lib/watershed').Watershed;

var shed = new Watershed();
var wskey = shed.generateKey();
var options = {
	port: 9554,
	hostname: '127.0.0.1',
	headers: {
		'connection': 'upgrade',
		'upgrade': 'websocket',
		'Sec-WebSocket-Key': wskey
	}
};
var req = http.request(options);
req.end();
req.on('upgrade', function (res, socket, head) {
	socket.setNoDelay(true);
	var wsc = shed.connect(res, socket, head, wskey);
	wsc.on('text', function (text) {
		console.log('received text: %s', text);
		wsc.end('thank you and good night');
	});
	wsc.on('end', function (code, reason) {
		console.log('end! (%s: %s)', code, reason ? reason : '<null>');
	});
	wsc.send('Hi there!');
});


test/basic/server.js
#!/usr/bin/env node
/*
 * Copyright 2016 Joyent, Inc.
 */

var http = require('http');
var Watershed = require('../../lib/watershed').Watershed;


var shed = new Watershed();



var srv = http.createServer();
srv.listen(9554);
srv.on('upgrade', function (req, socket, head) {
	var wsc;
	try {
		wsc = shed.accept(req, socket, head);
	} catch (ex) {
		console.error('error: ' + ex.message);
		socket.end();
		return;
	}
	wsc.on('text', function (text) {
		console.log('received text: ' + text);
	});
	wsc.on('end', function (code, reason) {
		console.log('end! (%s: %s)', code, reason);
	});
	wsc.send('hi from the server');
});


lib/watershed.js
/*
 * Watershed:
 *   An implementation of RFC6455 (The WebSocket Protocol)
 *
 * Copyright (c) 2017, Joyent, Inc.
 */

var crypto = require('crypto');
var assert = require('assert');
var EventEmitter = require('events').EventEmitter;
var util = require('util');
var stream = require('stream');
var dtrace = require('./dtrace');

/*
 * So that we can straddle v0.8.0 (Old Streams) and v0.10.0 (New Streams)
 * for some time, we will conditionally wrap old streams in the new
 * stream compatibility module.
 */
var WRAP_SOCKETS = false;
if (!stream.Readable) {
	WRAP_SOCKETS = true;
	stream = require('readable-stream');
}

/*
 * Symbolic Constants from the RFC:
 */
var MAGIC_WEBSOCKET_UUID = '258EAFA5-E914-47DA-95CA-C5AB0DC85B11';
var NONCE_LENGTH = 16;
var OPCODE = {
	CONT: 0x0,
	TEXT: 0x1,
	BINARY: 0x2,
	CLOSE: 0x8,
	PING: 0x9,
	PONG: 0xA
};
var CLOSECODE = {
	NORMAL: 1000,
	GOING_AWAY: 1001,
	PROTOCOL_ERROR: 1002,
	UNACCEPTABLE: 1003,
	MALFORMED: 1007,
	POLICY_VIOLATION: 1008,
	TOO_BIG: 1009,
	MISSING_EXTENSION: 1010,
	UNEXPECTED_ERROR: 1011
};

function
_sha1(str)
{
	var hash = crypto.createHash('sha1');
	hash.update(str);

	return (hash.digest('base64'));
}


function
_generateResponse(wskey, proto)
{
	var wsaccept = _sha1(wskey + MAGIC_WEBSOCKET_UUID);
	var lines = [
		'HTTP/1.1 101 The Watershed Moment',
		'Upgrade: websocket',
		'Connection: Upgrade',
		'Sec-WebSocket-Accept: ' + wsaccept
	];
	if (proto !== undefined) {
		lines.push('Sec-WebSocket-Protocol: ' + proto);
	}
	return (lines.join('\r\n') + '\r\n\r\n');
}

function
_wrapSocket(socket)
{
	var ret = socket;
	if (WRAP_SOCKETS) {
		ret = new stream.Readable();
		ret.wrap(socket);

		/*
		 * It seems like wrap does not unpause a stream,
		 * so force the issue:
		 */
		socket.resume();
	}
	return (ret);
}

function
_findCloseCode(code)
{
	var keys = Object.keys(CLOSECODE);
	for (var i = 0;	i < keys.length; i++) {
		var key = keys[i];
		if (CLOSECODE[key] === code)
			return (key);
	}
	return (null);
}

function
Watershed()
{
}

/*
 * Public:  Watershed.generateKey()
 *
 * Returns a random, Base64-encoded 16-byte value suitable for use as the
 * Sec-WebSocket-Key header on an Upgrade request.
 */
Watershed.prototype.generateKey = function
generateKey()
{
	return (crypto.randomBytes(NONCE_LENGTH).toString('base64'));
};

/*
 * Public:  Watershed.accept(http.ServerRequest, net.Socket, Buffer)
 *
 * Responds to a client's request to Upgrade to WebSockets and returns a
 * WatershedConnection/EventEmitter.  The EventEmitter emits the following
 * events:
 *
 *    'error':  there was an error while handling the connection.
 *    'end':    the WebSocket connection has ended.
 *
 *    'text':   a TEXT frame arrived; (parameter will be a String.)
 *    'binary': a BINARY frame arrived; (parameter will be a Buffer.)
 *    'ping':   a PING frame arrived; (parameter will be a nonce Buffer.)
 *    'pong':   a PONG frame arrived; (parameter will be a nonce Buffer.)
 */
Watershed.prototype.accept = function
accept(req, socket, head, detached, protocols)
{
	var remote = socket.remoteAddress + ':' + socket.remotePort;
	var local = socket.localAddress + ':' + socket.localPort;
	socket = _wrapSocket(socket);
	/*
	 * Return any potential parse overrun back to the
	 * front of the stream:
	 */
	if (head && head.length > 0)
		socket.unshift(head);

	/*
	 * Check for the requisite headers in the Upgrade request:
	 */
	var upgrade = req.headers['upgrade'];
	if (!upgrade || upgrade.toLowerCase() !== 'websocket')
		throw (new Error('Missing Upgrade Header'));
	var wskey = req.headers['sec-websocket-key'];
	if (!wskey)
		throw (new Error('Missing Sec-WebSocket-Key Header'));
	var wsver = req.headers['sec-websocket-version'];
	if (wsver && wsver !== '13')
		throw (new Error('Unsupported Sec-WebSocket-Version'));

	var proto;
	var supported = {};
	var protoHead = req.headers['sec-websocket-protocol'];
	if (protoHead && protocols) {
		protocols.forEach(function (pr) {
			supported[pr] = true;
		});
		var protosWanted = protoHead.split(/, */g);
		for (var i = 0; i < protosWanted.length; ++i) {
			if (supported[protosWanted[i]]) {
				proto = protosWanted[i];
				break;
			}
		}
		if (proto === undefined) {
			throw (new Error('Client and server have no matching ' +
			    'sub-protocols'));
		}
	} else if (protoHead) {
		throw (new Error('Client requested a sub-protocol but no ' +
		    'supported sub-protocols were provided to accept()'));
	}

	/*
	 * Write the response that lets the client know we've accepted the
	 * Upgrade to WebSockets:
	 */
	socket.write(_generateResponse(wskey, proto));

	if (detached === true) {
		/*
		 * The user just wants the Socket.
		 */
		return (socket);
	}

	var options = {
		remoteMustMask: true,
		localShouldMask: false,
		type: 'accept',
		remote: remote,
		local: local,
		protocol: proto
	};
	return (new WatershedConnection(options, socket));
};

/*
 * Public:  Watershed.connect(http.ClientResponse, net.Socket, Buffer, String)
 *
 * Attaches a new client-side WatershedConnection to this presently Upgraded
 * socket.  Emits the same events as the object returned by accept().
 */
Watershed.prototype.connect = function
connect(res, socket, head, wskey, detached)
{
	var remote = socket.remoteAddress + ':' + socket.remotePort;
	var local = socket.localAddress + ':' + socket.localPort;
	socket = _wrapSocket(socket);
	/*
	 * Return any potential parse overrun back to the
	 * front of the stream:
	 */
	if (head && head.length > 0)
		socket.unshift(head);

	/*
	 * Check for the requisite headers in the Upgrade response:
	 */
	if (res.statusCode !== 101)
		throw (new Error('Invalid Upgrade status code'));
	var connection = res.headers['connection'];
	if (!connection || connection.toLowerCase() !== 'upgrade')
		throw (new Error('Missing Connection Header'));
	var upgrade = res.headers['upgrade'];
	if (!upgrade || upgrade.toLowerCase() !== 'websocket')
		throw (new Error('Missing Upgrade Header'));
	var wsaccept = res.headers['sec-websocket-accept'];
	if (!wsaccept || wsaccept !== _sha1(wskey + MAGIC_WEBSOCKET_UUID))
		throw (new Error('Missing Sec-WebSocket-Accept Header'));
	var wsver = res.headers['sec-websocket-version'];
	if (wsver && wsver !== '13')
		throw (new Error('Unsupported Sec-WebSocket-Version'));
	var proto = res.headers['sec-websocket-protocol'];

	if (detached === true) {
		/*
		 * The user just wants the Socket.
		 */
		return (socket);
	}

	var options = {
		remoteMustMask: false,
		localShouldMask: true,
		type: 'connect',
		remote: remote,
		local: local,
		protocol: proto
	};
	return (new WatershedConnection(options, socket));
};

function
WatershedConnection(options, socket)
{
	var self = this;
	EventEmitter.call(this);

	self._id = dtrace.nextId();

	self._data = new Buffer(0);
	self._stats = {
		receivedFrames: 0,
		sentFrames: 0
	};

	/*
	 * We only want to write at most _one_ CLOSE frame, or 'end' event.
	 */
	self._close_written = false;
	self._close_received = false;
	self._end_emitted = false;

	self._close_code = null;
	self._close_reason = null;

	self._options = options;
	self._socket = socket;
	self._remote = options.remote;
	self._local = options.local;
	self._proto = null;
	if (options.protocol)
		self._proto = options.protocol;

	self._check_for_http_header = true;

	/*
	 * DTrace start/end probes:
	 */
	dtrace._watershed_probes['start'].fire(function () {
		return ([
			self._id,
			self._remote,
			self._local,
			options.type
		]);
	});
	self.on('end', function (code, reason) {
		dtrace._watershed_probes['end'].fire(function () {
			return ([
				self._id,
				self._remote,
				self._local,
				self._close_code,
				self._close_reason
			]);
		});
	});

	/*
	 * XXX The New Stream interface appears, in v0.10.10, to deliver any
	 * data in the buffer (e.g. a banner from the server, etc)
	 * _immediately_ when we add our first 'readable' handler, before we
	 * have a chance to return from the constructor.  Work around this
	 * for now.
	 */
	self._outofconstructor = false;
	process.nextTick(function () {
		self._outofconstructor = true;
		self._ws_readFromSocket();
	});
	self._socket.on('readable', function () {
		if (!self._outofconstructor)
			return;
		self._ws_readFromSocket();
	});
	self._socket.on('end', function () {
		/*
		 * If we did not receive a CLOSE frame, then the connection was
		 * terminated prematurely.
		 */
		if (!self._close_received)
			self.emit('connectionReset');

		if (self._end_emitted)
			return;
		self._end_emitted = true;

		self.emit('end', self._close_code, self._close_reason);
	});
	self._socket.on('error', function (err) {
		if (self._end_emitted)
			return;
		self._end_emitted = true;

		/*
		 * Unfortunately, in the case of a write-after-end error there
		 * is no error code set. In this case we check that the error
		 * message property is equal to the string 'write after end' as
		 * it is specified in the node runtime.
		 */
		if (err.code === 'ECONNRESET' ||
		    err.code === 'EPIPE' ||
		    err.message === 'write after end') {
			/*
			 * Treat end-of-stream errors as merely an end
			 * of stream.  If we received a CLOSE frame, it
			 * was a graceful end.  If we did not, it was not.
			 */
			if (!self._close_received)
				self.emit('connectionReset');

			self.emit('end', self._close_code, self._close_reason);
			return;
		}

		self.emit('error', err);
		self.emit('end');
	});
}
util.inherits(WatershedConnection, EventEmitter);

/*
 * Public: WatershedConnection.getProtocol()
 *
 * Returns the negotiated sub-protocol chosen during accept().
 */
WatershedConnection.prototype.getProtocol = function
getProtocol()
{
	return (this._proto);
};

/*
 * Public: WatershedConnection.end(reason)
 *
 * Send a close frame to the remote end of the connection, with an optional
 * reason string.
 */
WatershedConnection.prototype.end = function
end(reason)
{
	if (this._close_written)
		return;
	this._close_written = true;

	this._ws_writeClose(CLOSECODE.NORMAL, reason);
};

/*
 * Public: WatershedConnection.destroy()
 *
 * Immediately destroy the underlying socket, without sending a CLOSE
 * frame.
 */
WatershedConnection.prototype.destroy = function
destroy()
{
	if (this._socket !== null) {
		this._socket.removeAllListeners();
		this._socket.destroy();
		this._socket = null;
	}
	if (!this._end_emitted) {
		this.emit('end', this._close_code, this._close_reason);
		this._end_emitted = true;
	}
};

/*
 * Public: WatershedConnection.send(string)
 *           ^-- send a TEXT frame with this UTF-8 string
 *         WatershedConnction.send(Buffer)
 *           ^-- send a BINARY frame with this Buffer
 */
WatershedConnection.prototype.send = function
send(data)
{
	assert(typeof (data) === 'string' || Buffer.isBuffer(data));

	if (Buffer.isBuffer(data)) {
		this._ws_writeBinary(data);
	} else {
		this._ws_writeText(data);
	}
};

WatershedConnection.prototype._ws_readFromSocket = function
_ws_readFromSocket()
{
	/*
	 * Read and process all frames we have fully received:
	 */
	while (!this._end_emitted) {
		if (this._ws_readFrame()) {
			this._stats.receivedFrames++;
		} else {
			break;
		}
	}
};

/*
 * Ensure that at least 'len' bytes are contiguous in the data buffer.
 *
 * Return false if there is insufficient data in the chain, otherwise return
 * true.
 */
WatershedConnection.prototype._ws_pullup = function
_ws_pullup(len)
{
	var self = this;
	var buf;

	/*
	 * We already have sufficient data, so do nothing.
	 */
	if (this._data.length >= len)
		return (true);

	/*
	 * Try and read data from the socket:
	 */
	buf = this._socket.read(len !== null ? len - this._data.length : null);
	if (buf === null)
		return (false);

	dtrace._watershed_probes['read-buffer'].fire(function () {
		return ([
			self._id,
			self._remote,
			self._local,
			buf.toString('binary'),
			buf.length
		]);
	});

	this._data = Buffer.concat([ this._data, buf ],
	    this._data.length + buf.length);

	return (true);
};

WatershedConnection.prototype._ws_writeBinary = function
_ws_writeBinary(buffer)
{
	var self = this;
	assert(Buffer.isBuffer(buffer));
	dtrace._watershed_probes['send-binary'].fire(function () {
		return ([
			self._id,
			self._remote,
			self._local,
			buffer.toString('binary'),
			buffer.length
		]);
	});
	this._ws_writeFrameCommon(OPCODE.BINARY, buffer);
};

WatershedConnection.prototype._ws_writeText = function
_ws_writeText(text)
{
	var self = this;
	assert(typeof (text) === 'string');
	dtrace._watershed_probes['send-text'].fire(function () {
		return ([
			self._id,
			self._remote,
			self._local,
			text,
			text.length
		]);
	});
	this._ws_writeFrameCommon(OPCODE.TEXT, new Buffer(text, 'utf8'));
};

WatershedConnection.prototype._ws_writeClose = function
_ws_writeClose(code, reason)
{
	var self = this;
	var buf;
	assert(code >= 1000);
	if (reason) {
		assert(typeof (reason) === 'string');
		buf = new Buffer(2 + Buffer.byteLength(reason, 'utf8'));
		buf.write(reason, 2);
	} else {
		buf = new Buffer(2);
	}
	buf.writeUInt16BE(code, 0);
	dtrace._watershed_probes['send-close'].fire(function () {
		return ([
			self._id,
			self._remote,
			self._local,
			buf.toString('binary'),
			buf.length
		]);
	});
	this._ws_writeFrameCommon(OPCODE.CLOSE, buf);
};

WatershedConnection.prototype._ws_writePing = function
_ws_writePing(nonce)
{
	this._ws_writeFrameCommon(OPCODE.PING, nonce);
};

WatershedConnection.prototype._ws_writePong = function
_ws_writePong(nonce)
{
	this._ws_writeFrameCommon(OPCODE.PONG, nonce);
};

WatershedConnection.prototype._ws_writeFrameCommon = function
_ws_writeFrameCommon(opcode, data)
{
	var maskbuf = null;
	var hdr;
	var obj = {
		fin: true,
		opcode: opcode
	};

	assert(Buffer.isBuffer(data));

	/*
	 * According to the RFC, the client MUST mask their outgoing frames.
	 */
	if (this._options.localShouldMask) {
		maskbuf = crypto.randomBytes(4);
		for (var j = 0; j < data.length; j++) {
			data[j] = data[j] ^ maskbuf[j % maskbuf.length];
		}
	}

	/*
	 * Construct the type of payload length we need:
	 */
	if (data.length <= 125) {
		hdr = new Buffer(2);
		obj.len0 = data.length;
	} else if (data.length <= 0xffff) {
		hdr = new Buffer(2 + 2);
		obj.len0 = 126;
		hdr.writeUInt16BE(data.length, 2);
	} else if (data.length <= 0xffffffff) {
		hdr = new Buffer(2 + 8);
		obj.len0 = 127;
		hdr.writeUInt32BE(0, 2);
		hdr.writeUInt32BE(data.length, 6);
	} else {
		throw (new Error('Frame payload must have length less ' +
		    'than 32-bits'));
	}

	/*
	 * Construct the common (first) two bytes of the header:
	 */
	var w0 = obj.fin ? (1 << 15) : 0;
	w0 |= (obj.opcode << 8) & 0x0f00;
	w0 |= obj.len0 & 0x007f;
	w0 |= maskbuf !== null ? (1 << 7) : 0;
	hdr.writeUInt16BE(w0, 0);

	/*
	 * Write the data:
	 */
	this._socket.write(hdr);
	if (maskbuf !== null)
		this._socket.write(maskbuf);
	this._socket.write(data);
	this._stats.sentFrames++;
};

WatershedConnection.prototype._ws_readFrame = function
_ws_readFrame()
{
	var self = this;
	var pos = 0;

	/*
	 * Read the common (first) two bytes of the header:
	 */
	if (!this._ws_pullup(pos + 2))
		return (false);
	var w0 = this._data.readUInt16BE(pos);
	pos += 2;

	if (this._check_for_http_header) {
		/*
		 * XXX There have been some horrible Streams/HTTP bugs that
		 * mean we get to see the HTTP headers on the front of the
		 * stream, or at a random point within the stream, even
		 * though it _should_ have been eaten by the HTTP parser.
		 *
		 * This check attempts to detect such malfeasance.
		 */
		if (this._data.toString('utf8', 0, 2) === 'HT') {
			throw (new Error('POSSIBLE NODE/STREAMS BUG'));
		}
	}

	/*
	 * Break the header bytes out into fields:
	 */
	var obj = {
		fin: !!(w0 & (1 << 15)),
		opcode: (w0 & 0x0f00) >> 8,
		mask: !!(w0 & (1 << 7)),
		len0: w0 & 0x007f,
		maskbytes: []
	};

	if (this._options.remoteMustMask && !obj.mask) {
		/*
		 * According to the RFC, the client MUST currently mask their
		 * frames.
		 */
		this._end_emitted = true;
		this.emit('error', new Error('Client did not Mask according ' +
		    'to the RFC.'));
		this.emit('end');
		this._socket.end();
		return (false);
	}

	/*
	 * XXX We should handle multi-part messages:
	 */
	if (!obj.fin) {
		this.end();
		return (false);
	}

	/*
	 * Determine the payload length; this may be in the common bytes, or in
	 * an additional field.
	 */
	assert(obj.len0 >= 0 && obj.len0 <= 127);
	if (obj.len0 <= 125) {
		obj.len = obj.len0;
	} else if (obj.len0 === 126) {
		if (!this._ws_pullup(pos + 2))
			return (false);
		obj.len = this._data.readUInt16BE(pos);
		pos += 2;
	} else { /* obj.len === 127 */
		if (!this._ws_pullup(pos + 4))
			return (false);
		obj.len = this._data.readUInt32BE(pos);
		pos += 4;
		/*
		 * XXX We cannot usefully use a 64-bit value, so make sure the
		 * upper 32 bits are zero for now.
		 */
		if (obj.len !== 0) {
			this._end_emitted = true;
			this.emit('error', new Error('Client tried to send ' +
			    'too long a frame.'));
			this.emit('end');
			this._socket.end();
			return (false);
		}

		if (!this._ws_pullup(pos + 4))
			return (false);
		obj.len = this._data.readUInt32BE(pos);
		pos += 4;
	}

	/*
	 * Read the remote connection's mask key:
	 */
	if (obj.mask) {
		if (!this._ws_pullup(pos + 4))
			return (false);
		for (var i = 0; i < 4; i++) {
			obj.maskbytes.push(this._data.readUInt8(pos));
			pos++;
		}
	}

	/*
	 * Load the payload:
	 */
	if (!this._ws_pullup(pos + obj.len))
		return (false);
	obj.payload = this._data.slice(pos, pos + obj.len);
	pos += obj.len;

	/*
	 * If the remote connection masked their payload, unmask it:
	 */
	if (obj.mask) {
		for (var k = 0; k < obj.payload.length; k++) {
			obj.payload[k] = obj.payload[k] ^ obj.maskbytes[k % 4];
		}
	}

	if (obj.opcode === OPCODE.CLOSE)
		this._close_received = true;

	/*
	 * Emit events for the frame:
	 */
	if (obj.opcode === OPCODE.TEXT) {
		var stringOut = obj.payload.toString('utf8');
		dtrace._watershed_probes['recv-text'].fire(function () {
			return ([
				self._id,
				self._remote,
				self._local,
				stringOut,
				stringOut.length
			]);
		});
		this.emit('text', stringOut);
	} else if (obj.opcode === OPCODE.BINARY) {
		dtrace._watershed_probes['recv-binary'].fire(function () {
			return ([
				self._id,
				self._remote,
				self._local,
				obj.payload.toString('binary'),
				obj.payload.length
			]);
		});
		this.emit('binary', obj.payload);
	} else if (obj.opcode === OPCODE.PING) {
		this.emit('ping', obj.payload);
		/*
		 * XXX We should probably let the user do this for themselves:
		 */
		this._ws_writePong(obj.payload);
	} else if (obj.opcode === OPCODE.PONG) {
		this.emit('pong', obj.payload);
	} else if (obj.opcode === OPCODE.CLOSE) {
		/*
		 * We've received a CLOSE frame, either as a result of a
		 * remote-initiated CLOSE, or in response to a CLOSE frame we
		 * sent.  In the former case, the RFC dictates that we respond
		 * in kind; otherwise close the socket.
		 */
		dtrace._watershed_probes['recv-close'].fire(function () {
			return ([
				self._id,
				self._remote,
				self._local,
				obj.payload.toString('binary'),
				obj.payload.length
			]);
		});
		if (obj.payload.length >= 2) {
			this._close_code = _findCloseCode(
			    obj.payload.readUInt16BE(0));
			this._close_reason = obj.payload.toString('utf8', 2);
		}
		this.end();
		this._socket.end();
	}

	/*
	 * Turf this frame out of the front of the chain.
	 */
	this._data = this._data.slice(pos);
	return (true);
};



module.exports = {
	Watershed: Watershed
};

/* vim: set sts=8 ts=8 sw=8 tw=80 noet: */
